src/cloudai/_core/base_job.py

```diff
@@ -14,40 +14,121 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Union
+
+from .system import System
 from .test import Test
 
 
-class BaseJob:
+class BaseJob(ABC):
     """
     Base class for representing a job created by executing a test.
 
     Attributes
-        id (int): The unique identifier of the job.
+        mode (str): The mode of the job (e.g., 'run', 'dry-run').
+        system (System): The system in which the job is running.
         test (Test): The test instance associated with this job.
-        output_path (str): The path where the job's output is stored.
+        output_path (Path): The path where the job's output is stored.
         terminated_by_dependency (bool): Flag to indicate if the job was terminated due to a dependency.
     """
 
-    def __init__(self, job_id: int, test: Test, output_path: str):
+    def __init__(self, mode: str, system: System, test: Test, output_path: Path):
         """
         Initialize a BaseJob instance.
 
         Args:
-            job_id (int): The unique identifier of the job.
-            output_path (str): The path where the job's output is stored.
+            mode (str): The mode of the job (e.g., 'run', 'dry-run').
+            system (System): The system in which the job is running.
             test (Test): The test instance associated with the job.
+            output_path (Path): The path where the job's output is stored.
         """
-        self.id = job_id
+        self.mode = mode
+        self.system = system
         self.test = test
         self.output_path = output_path
         self.terminated_by_dependency = False
 
-    def increment_iteration(self):
+    def get_mode(self) -> str:
```

Comment:
In python we usually don't need getters and setters, because all properties are public. It looks too complicated, especially in this very simplistic form where you just return the value as it is.

In case we ever need to change it, python still has us covered, because we will just create a `@property` that will hide logic and won't require any change on a caller's side.

I suggest we remove all these `get_*` methods to keep code simply. And we can keep it also expressive, especially with converting `is_running` and `is_completed` into properties too. Caller's code will look more pythonic.

Reply:
> In python we usually don't need getters and setters, because all properties are public. It looks too complicated, especially in this very simplistic form where you just return the value as it is.

That makes sense, and this approach aligns with what I generally follow. I will make the necessary updates and avoid adding getters or setters.

> And we can keep it also expressive, especially with converting is_running and is_completed into properties too. 

Does this mean we should convert is_running into a property like job.status?

Reply:
> Does this mean we should convert is_running into a property like job.status?

It make sense to me. So if you don't have any objections, let's go with this approach and use properties.

Reply:
Sure, I will do it. BTW, does this apply to get_id() as well? Actually, I took this approach to maintain consistency after adding get_id(). get_id() returns a unique job ID for a given job, which is an abstract method later inherited by other job classes. Do you think we should take the property approach for this as well?

Reply:
Make sense to me. But with `id` name it might be slightly different, because it is a keyword in Python. So maybe we rename it to `jobid` or something like that.

Reply:
Please find the updated code.